File: src/core/RuleEngine.js
/**
 * RuleEngine — 规则引擎
 * 处理确定性的 Optimizer 决策 & Review Gate 审批
 * OpenAutoGrowth Core
 */
export class RuleEngine {
    constructor() {
        // 内置规则集（匹配 docs/architecture/05-rule-engine.md）
        this.rules = this._loadDefaultRules();
        this.cooldownTracker = new Map(); // Map<"campaignId:ruleId", lastFiredTimestamp>
    }

    /**
     * 执行规则引擎，返回 OptAction[]
     * @param {object} context - { metrics, campaign, ab_test, anomalies, kpi }
     * @param {string} campaignId
     * @returns {Array} actions
     */
    evaluate(context, campaignId) {
        const sortedRules = [...this.rules].sort((a, b) => a.priority - b.priority);
        const actions = [];

        for (const rule of sortedRules) {
            if (!rule.enabled) continue;
            if (this._isOnCooldown(campaignId, rule.id, rule.cooldown_ms)) continue;

            if (this._evaluateCondition(rule.condition, context)) {
                console.log(`[RuleEngine] ✅ Rule fired: ${rule.name}`);
                actions.push({ rule_id: rule.id, ...rule.action });
                this._recordFire(campaignId, rule.id);
            }
        }

        // 去重：同类型 action 只保留优先级最高的
        return this._deduplicateActions(actions);
    }

    /**
     * 递归评估 Condition 表达式
     */
    _evaluateCondition(condition, context) {
        if ('field' in condition) {
            return this._evaluatePredicate(condition, context);
        }

        const results = condition.children.map(c => this._evaluateCondition(c, context));
        switch (condition.operator) {
            case 'AND': return results.every(Boolean);
            case 'OR':  return results.some(Boolean);
            case 'NOT': return !results[0];
            default:    return false;
        }
    }

    _evaluatePredicate({ field, op, value }, context) {
        const actual = this._getFieldValue(field, context);
        if (actual === undefined || actual === null) return false;

        switch (op) {
            case '>':    return actual > value;
            case '<':    return actual < value;
            case '>=':   return actual >= value;
            case '<=':   return actual <= value;
            case '==':   return actual === value;
            case 'IN':   return Array.isArray(value) && value.includes(actual);
            case 'NOT_IN': return Array.isArray(value) && !value.includes(actual);
            default:     return false;
        }
    }

    _getFieldValue(fieldPath, context) {
        return fieldPath.split('.').reduce((obj, key) => obj?.[key], context);
    }

    _isOnCooldown(campaignId, ruleId, cooldownMs) {
        const key = `${campaignId}:${ruleId}`;
        const lastFired = this.cooldownTracker.get(key);
        return lastFired && (Date.now() - lastFired) < cooldownMs;
    }

    _recordFire(campaignId, ruleId) {
        this.cooldownTracker.set(`${campaignId}:${ruleId}`, Date.now());
    }

    _deduplicateActions(actions) {
        const seen = new Set();
        return actions.filter(a => {
            const key = a.type + (a.params?.target || '');
            if (seen.has(key)) return false;
            seen.add(key);
            return true;
        });
    }

    _loadDefaultRules() {
        return [
            {
                id: 'R005', name: '异常告警 - 人工介入',
                priority: 0, enabled: true, cooldown_ms: 3600000,
                condition: {
                    operator: 'OR', children: [
                        { field: 'anomalies.cpm_surge_pct',  op: '>=', value: 0.5 },
                        { field: 'anomalies.ctr_drop_pct',   op: '>=', value: 0.4 },
                    ]
                },
                action: { type: 'ALERT_HUMAN', params: { pause_campaign: true } }
            },
            {
                id: 'R001', name: 'A/B 裁决 - 置信度达标',
                priority: 1, enabled: true, cooldown_ms: 86400000,
                condition: {
                    operator: 'AND', children: [
                        { field: 'ab_test.winner_confidence',    op: '>=', value: 0.95 },
                        { field: 'ab_test.min_sample_size_met', op: '==', value: true },
                    ]
                },
                action: { type: 'PAUSE_VARIANT', params: { target: 'loser_variant_id' } }
            },
            {
                id: 'R002', name: 'CTR 严重低于基准',
                priority: 2, enabled: true, cooldown_ms: 43200000,
                condition: {
                    operator: 'AND', children: [
                        { field: 'metrics.ctr_below_baseline_70pct', op: '==', value: true },
                        { field: 'metrics.impressions',              op: '>=', value: 1000 },
                    ]
                },
                action: { type: 'TRIGGER_REWRITE', params: { agent: 'ContentGen', urgency: 'HIGH' } }
            },
            {
                id: 'R003', name: 'ROAS 不达标 - 预算收缩',
                priority: 3, enabled: true, cooldown_ms: 21600000,
                condition: {
                    operator: 'AND', children: [
                        { field: 'metrics.roas',         op: '<',  value: 2.0 },
                        { field: 'campaign.loop_count',  op: '<',  value: 5 },
                    ]
                },
                action: { type: 'REALLOCATE_BUDGET', params: { shrink_pct: 0.3, reallocate_to: 'best_channel' } }
            },
            {
                id: 'R004', name: 'ROI 超额 - 提速扩量',
                priority: 4, enabled: true, cooldown_ms: 43200000,
                condition: { field: 'metrics.roas_exceeds_target_120pct', op: '==', value: true },
                action: { type: 'SCALE_BUDGET', params: { scale_pct: 0.2 } }
            },
        ];
    }
}

File: backend/app/core/rule_engine.py
"""
RuleEngine — deterministic optimizer decision engine.
Python port of src/core/RuleEngine.js with full DSL condition evaluation.
Rules: R001-R005 (see docs/architecture/05-rule-engine.md)
"""
import time
from typing import Any

import structlog

logger = structlog.get_logger(__name__)


# ── Default rule set (mirrors RuleEngine.js) ──────────────────────────────────

DEFAULT_RULES = [
    {
        "id": "R001",
        "name": "低 CTR 暂停",
        "priority": 10,
        "cooldown_ms": 3_600_000,
        "enabled": True,
        "condition": {
            "operator": "AND",
            "children": [
                {"field": "metrics.ctr",  "op": "<",  "value": 0.01},
                {"field": "metrics.roas", "op": "<",  "value": 1.5},
            ],
        },
        "action": {"type": "PAUSE_AD_GROUP", "params": {"reason": "low_ctr"}},
    },
    {
        "id": "R002",
        "name": "高 ROAS 加量",
        "priority": 20,
        "cooldown_ms": 7_200_000,
        "enabled": True,
        "condition": {
            "operator": "AND",
            "children": [
                {"field": "metrics.roas", "op": ">",  "value": 3.5},
                {"field": "metrics.ctr",  "op": ">",  "value": 0.025},
            ],
        },
        "action": {"type": "INCREASE_BUDGET", "params": {"pct": 0.20}},
    },
    {
        "id": "R003",
        "name": "A/B 胜者确认",
        "priority": 30,
        "cooldown_ms": 86_400_000,
        "enabled": True,
        "condition": {
            "operator": "AND",
            "children": [
                {"field": "metrics.clicks", "op": ">", "value": 500},
            ],
        },
        "action": {"type": "PAUSE_LOSING_VARIANTS", "params": {}},
    },
    {
        "id": "R004",
        "name": "异常 ROAS 告警",
        "priority": 5,
        "cooldown_ms": 1_800_000,
        "enabled": True,
        "condition": {
            "operator": "OR",
            "children": [
                {"field": "metrics.roas", "op": "<",  "value": 0.5},
                {"field": "metrics.roas", "op": ">",  "value": 15.0},
            ],
        },
        "action": {"type": "ALERT", "params": {"severity": "HIGH", "metric": "roas"}},
    },
    {
        "id": "R005",
        "name": "超预算熔断",
        "priority": 1,
        "cooldown_ms": 0,
        "enabled": True,
        "condition": {"field": "metrics.spend_ratio", "op": ">", "value": 0.95},
        "action": {"type": "PAUSE_CAMPAIGN", "params": {"reason": "budget_exhausted"}},
    },
]


class RuleEngine:
    def __init__(self, rules: list[dict] | None = None):
        self.rules = sorted(rules or DEFAULT_RULES, key=lambda r: r["priority"])
        self._cooldown: dict[str, float] = {}   # "campaignId:ruleId" → last_fired_ms

    def evaluate(self, context: dict, campaign_id: str) -> list[dict]:
        """
        Evaluate all enabled rules against context.
        Returns deduplicated list of OptAction dicts.
        """
        actions = []
        for rule in self.rules:
            if not rule.get("enabled", True):
                continue
            if self._on_cooldown(campaign_id, rule["id"], rule.get("cooldown_ms", 0)):
                continue
            if self._eval_condition(rule["condition"], context):
                logger.info("rule_fired", rule_id=rule["id"], name=rule["name"])
                actions.append({"rule_id": rule["id"], **rule["action"]})
                self._record_fire(campaign_id, rule["id"])

        return self._dedup(actions)

    # ── Condition evaluation ──────────────────────────────────────

    def _eval_condition(self, condition: dict, context: dict) -> bool:
        if "field" in condition:
            return self._eval_predicate(condition, context)

        operator = condition.get("operator", "AND")
        children = condition.get("children", [])
        results = [self._eval_condition(c, context) for c in children]

        if operator == "AND":
            return all(results)
        if operator == "OR":
            return any(results)
        if operator == "NOT":
            return not results[0] if results else False
        return False

    def _eval_predicate(self, pred: dict, context: dict) -> bool:
        field = pred["field"]
        op    = pred["op"]
        value = pred["value"]

        actual = self._get_field(field, context)
        if actual is None:
            return False

        match op:
            case ">":  return actual > value
            case ">=": return actual >= value
            case "<":  return actual < value
            case "<=": return actual <= value
            case "==": return actual == value
            case "!=": return actual != value
            case _:    return False

    def _get_field(self, field: str, context: dict) -> Any:
        """Dot-notation field access: 'metrics.ctr' → context['metrics']['ctr']"""
        parts = field.split(".")
        node = context
        for part in parts:
            if not isinstance(node, dict) or part not in node:
                return None
            node = node[part]
        return node

    # ── Cooldown tracking ─────────────────────────────────────────

    def _on_cooldown(self, campaign_id: str, rule_id: str, cooldown_ms: int) -> bool:
        if cooldown_ms <= 0:
            return False
        key = f"{campaign_id}:{rule_id}"
        last = self._cooldown.get(key, 0)
        return (time.time() * 1000 - last) < cooldown_ms

    def _record_fire(self, campaign_id: str, rule_id: str):
        self._cooldown[f"{campaign_id}:{rule_id}"] = time.time() * 1000

    # ── Deduplication ─────────────────────────────────────────────

    def _dedup(self, actions: list[dict]) -> list[dict]:
        """Keep only the highest-priority action per type."""
        seen: dict[str, dict] = {}
        for action in actions:
            t = action["type"]
            if t not in seen:
                seen[t] = action
        return list(seen.values())


# Global singleton
rule_engine = RuleEngine()
